src/main/scala/cromwell/CromwellCommandLine.scala

```diff
@@ -0,0 +1,104 @@
+package cromwell
+
+import java.nio.file.{Files, Path, Paths}
+
+import better.files._
+import cromwell.core.{ErrorOr, WorkflowSourceFiles}
+import cromwell.util.FileUtil._
+import lenthall.exception.MessageAggregation
+
+import scalaz._
+import Scalaz._
+import scala.util.{Failure, Success, Try}
+
+sealed abstract class CromwellCommandLine
+case object UsageAndExit extends CromwellCommandLine
+case object RunServer extends CromwellCommandLine
+final case class RunSingle(wdlPath: Path,
+                           sourceFiles: WorkflowSourceFiles,
+                           inputsPath: Option[Path],
+                           optionsPath: Option[Path],
+                           metadataPath: Option[Path]) extends CromwellCommandLine
+
+object CromwellCommandLine {
+  def apply(args: Seq[String]): CromwellCommandLine = {
+    args.headOption match {
+      case Some("server") if args.size == 1 => RunServer
+      case Some("run") if args.size >= 2 && args.size <= 5 => RunSingle(args.tail)
+      case _ => UsageAndExit
+    }
+  }
+}
+
+object RunSingle {
+  def apply(args: Seq[String]): RunSingle = {
+    val wdlPath = Paths.get(args.head)
+    val inputsPath = argPath(args, 1, Option(".inputs"), checkDefaultExists = false)
+    val optionsPath = argPath(args, 2, Option(".options"), checkDefaultExists = true)
+    val metadataPath = argPath(args, 3, None)
+
+    val wdl = readContent("WDL file", wdlPath)
+    val inputsJson = readJson("Inputs", inputsPath)
+    val optionsJson = readJson("Workflow Options", optionsPath)
+
+    (wdl |@| inputsJson |@| optionsJson) {
+      (w, i, o) =>
+        val workflowSourceFiles = WorkflowSourceFiles(w, i, o)
+        metadataPath foreach { m => if (!metadataPathIsWriteable(m)) throw new RuntimeException(s"Unable to write to metadata directory: $m")}
+        RunSingle(wdlPath, workflowSourceFiles, inputsPath, optionsPath, metadataPath)
+    } match {
+      case scalaz.Success(r) => r
+      case scalaz.Failure(nel) => throw new RuntimeException with MessageAggregation {
+        override def exceptionContext: String = "ERROR: Unable to run Cromwell:"
+        override def errorMessages: Traversable[String] = nel.list
+      }
+    }
```

Comment:
Disclaimer: didn't actually try this outside IntelliJ, but it does keep everything in `ErrorOr` until the final failure check:

```scala
    def writeableMetadataPath(path: Option[Path]): ErrorOr[Unit] = {
      path match {
        case Some(p) if !metadataPathIsWriteable(p) => s"Unable to write to metadata directory: $p".failureNel
        case otherwise => ().successNel
      }
    }

    val sourceFiles: ErrorOr[WorkflowSourceFiles] = (wdl |@| inputsJson |@| optionsJson) {
      WorkflowSourceFiles.apply
    }

    val runSingle = for {
      sources <- sourceFiles
      _ <- writeableMetadataPath(metadataPath)
    } yield RunSingle(wdlPath, sources, inputsPath, optionsPath, metadataPath)

    runSingle match {
      case scalaz.Success(r) => ...
    }
```